### PR TITLE
Issue 48 Design Database Schema for Messaging

### PR DESCRIPTION
**Pull request by MaksymYoung**
You can see that MongoDB connection works:
image
image
image
image
image

### CODE REVIEW

File: chat/management/commands/populate_messages.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            user = UserDocument.objects(email=u["email"]).first()
+            if not user:
+                user = UserDocument(
+                    email=u["email"],
+                    first_name=u["first_name"],
+                    last_name=u["last_name"],
+                    password="pass123",
+                    role=role_user,
+                    is_active=True,
+                ).save()
```
Comment (mehalyna) [inline chat/management/commands/populate_messages.py:40]: Wrap user creation in a try/except and log failures to help debug issues with user data or DB connectivity.

File: chat/management/commands/populate_messages.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            users.append(user)
```
Comment (mehalyna) [inline chat/management/commands/populate_messages.py:41]: Hardcoded password "pass123" is insecure. Consider using a secure password or making it configurable. Add a comment warning not to use this in production.
If your UserDocument hashes passwords, ensure this is respected; otherwise, users may have unusable credentials.

File: chat/management/commands/populate_messages.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        room = Room.objects(name="TestRoom").first()
```
Comment (mehalyna) [inline chat/management/commands/populate_messages.py:43]: If you want to allow multiple test rooms, make room name configurable via options.

File: chat/management/commands/populate_messages.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        if not room:
+            room = Room(name="TestRoom", participants=users).save()
```
Comment (mehalyna) [inline chat/management/commands/populate_messages.py:45]: If Room creation fails due to validation errors (e.g., duplicate participants, max participants), log or surface the error.

File: chat/management/commands/populate_messages.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+        for _ in range(10):
```
Comment (mehalyna) [inline chat/management/commands/populate_messages.py:47]: Make the number of messages configurable. Also, consider shuffling senders/messages to ensure coverage of all users and messages.

File: tests/chat/test_chat.py [hunk 6]
```diff
@@ -1,54 +1,75 @@
...
+from mongoengine import connect, disconnect
 from selenium import webdriver
 from selenium.webdriver.chrome.options import Options
 from selenium.webdriver.chrome.service import Service
 from selenium.webdriver.common.action_chains import ActionChains
 from selenium.webdriver.common.keys import Keys
 from webdriver_manager.chrome import ChromeDriverManager
+from users.documents import UserDocument, UserRoleDocument, UserRoleEnum
 
+os.environ['DJANGO_SETTINGS_MODULE'] = 'core.test_settings'
```
Comment (mehalyna) [inline tests/chat/test_chat.py:15]: Consider moving environment setup into a test-specific entrypoint or pytest fixture for clarity and isolation.

File: tests/chat/test_chat.py [hunk 7]
```diff
@@ -1,54 +1,75 @@
...
+            role.save()
+
+        cls.user = UserDocument(
+            email="testuser@example.com",
+            first_name="Test",
+            last_name="User",
+            password="testpass",
+            role=role
+        )
+        cls.user.save()
```
Comment (mehalyna) [inline tests/chat/test_chat.py:45]: The password is hardcoded and not hashed. If your UserDocument requires a hashed password, ensure this is done. Add a comment warning this is for test only.
Add a check or exception handling for duplicate users.
Add a teardown process to remove test users after tests run to avoid polluting the test DB.

File: tests/chat/test_chat.py
Comment (mehalyna) [inline tests/chat/test_chat.py:1]: - Add a test for edge cases: e.g., empty message, forbidden content, or exceeding message length.
- Consider marking these tests as integration/slow, especially if they use Selenium and external services.
- If tests are flaky, investigate using WebDriverWait and explicit expected conditions for more resilience.
- Ensure all created test data is cleaned up after tests to avoid polluting persistent or shared environments.
- Document all helper methods with parameter and return type hints for clarity and maintainability.

**Review by mehalyna — APPROVED** [review]
The failing job is due to multiple chat-related tests raising TimeoutError, indicating that certain messages are not being blocked or filtered as expected. The problematic tests are:
- test_empty_message
- test_exceeding_length_message
- test_forbidden_content_message
- test_message_not_visible_to_different_room
- test_message_visible_to_same_room

**Root Cause:**  
These tests rely on the chat system to block empty, excessively long, or forbidden-content messages and to ensure message isolation between chat rooms. The repeated TimeoutError suggests the backend is either allowing these messages through, or not updating the client UI as expected when a message is properly blocked.

**Solution:**
1. **Backend Message Validation:**  
   Ensure your backend chat handling logic validates and blocks:
   - Empty messages
   - Messages exceeding `MAX_MESSAGE_LENGTH`
   - Messages containing any word in `FORBIDDEN_WORDS_SET`
   - Messages sent to the wrong room or broadcast to the wrong recipients

   Example patch for Django Channels consumer:
   ```python
   # In your chat consumer
   from core.settings import FORBIDDEN_WORDS_SET
   MAX_MESSAGE_LENGTH = 1000  # Update if defined elsewhere

   def message_is_valid(message):
       if not message.strip():
           return False
       if len(message) > MAX_MESSAGE_LENGTH:
           return False
       if any(word in message for word in FORBIDDEN_WORDS_SET):
           return False
       return True

   async def receive(self, text_data):
       # ...parse message...
       if not message_is_valid(message):
           return  # Silently ignore invalid messages
       # Proceed to broadcast valid messages
   ```
   Adjust the import and validation to match your actual consumer structure.

2. **Frontend/UI Updates:**  
   If you expect the UI to show an error for invalid messages, ensure your JavaScript/WebSocket handler displays appropriate feedback and does not update the chat log for blocked messages.

3. **Test Environment Consistency:**  
   The test uses Selenium and expects the chat log to update only for valid messages. Make sure your application in test mode behaves exactly as in production regarding message validation.

**References:**
- [Chat test code](https://github.com/Project-Stage-Academy/UA-4148-bravo/blob/12b0d6ccef187d3d69e035ec974cc75e16065b9d/tests/chat/test_chat.py)
- [CI workflow definition](https://github.com/Project-Stage-Academy/UA-4148-bravo/blob/12b0d6ccef187d3d69e035ec974cc75e16065b9d/.github/workflows/ci-cd.yml)

Correcting your backend validation as shown should resolve the TimeoutError in these tests and allow the job to pass.

**Review by PaulBein — APPROVED** [review]
Great job!

**Review by DimaMakovskyi — APPROVED** [review]
GOOD JOB
